packages/flutter_tools/lib/src/base/deferred_component.dart

```diff
@@ -0,0 +1,291 @@
+  LoadingUnit({
+    @required this.id,
+    @required this.libraries,
+    this.path,
+  });
+
+  /// The unique loading unit id that is used to identify the loading unit within dart.
+  final int id;
+
+  /// A list of dart libraries that the loading unit contains.
+  final List<String> libraries;
+
+  /// The output path of the shared library .so file created by gen_snapshot.
+  ///
+  /// This value may be null when the loading unit is parsed from a
+  /// `deferred_components_golden.yaml` file, which does not store the path.
+  final String path;
+
+  /// Returns a human readable string representation of this LoadingUnit, ignoring
+  /// the [path] field. The [path] is not included as it is not relevant when the
+  @override
+  String toString() {
+    String out = '\nLoadingUnit $id\n  Libraries:';
+    for (final String lib in libraries) {
+      out += '\n  - $lib';
+    }
+    return out;
+  }
+
+  /// Returns true if the other loading unit has the same [id] and the same set of [libraries],
+  /// ignoring order.
+  bool equalsIgnoringPath(LoadingUnit other) {
+    return other.id == id && other.libraries.toSet().containsAll(libraries);
+  }
+
+  /// Parses the loading unit manifests from the [Environment.outputDir] of the latest
+  /// gen_snapshot/assemble run.
+  ///
+  /// This will read all existing loading units for every provided abi. If no abis are
+  /// provided, loading units for all abis will be parsed.
+  static List<LoadingUnit> parseGeneratedLoadingUnits(Environment env, {List<String> abis}) {
+    final List<LoadingUnit> loadingUnits = <LoadingUnit>[];
+    final List<FileSystemEntity> files = env.outputDir.listSync(recursive: true);
+    while (files.isNotEmpty) {
+      if (files.last is File) {
+        final File file = files.last as File;
+        // Determine if the abi is one we build.
+        bool matchingAbi = abis == null;
+        if (abis != null) {
+          for (final String abi in abis) {
+            if (file.parent.path.endsWith(abi)) {
+              matchingAbi = true;
+              break;
+            }
+          }
+        }
+        if (!file.path.endsWith('manifest.json') || !matchingAbi) {
+          files.removeLast();
+          continue;
+        }
+        loadingUnits.addAll(parseLoadingUnitManifest(file));
+      }
+      files.removeLast();
+    }
+    return loadingUnits;
+  }
+
+  /// Parses loading units from a single loading unit manifest json file.
+  ///
+  /// Returns an empty list if the manifestFile does not exist or is invalid.
+  static List<LoadingUnit> parseLoadingUnitManifest(File manifestFile) {
+    if (!manifestFile.existsSync()) {
+      return <LoadingUnit>[];
+    }
+    // Read gen_snapshot manifest
+    final String fileString = manifestFile.readAsStringSync();
+    Map<String, dynamic> manifest;
+    try {
+      manifest = jsonDecode(fileString) as Map<String, dynamic>;
+    } on FormatException catch (e) {
+      globals.printError(e.toString());
```

Comment:
Does this error need more context? something like "Loading unit manifest was invalid JSON, to fix ... \n $error"